narwhal/types/src/primary.rs

```diff
@@ -237,8 +237,11 @@ impl Header {
             }
         );
 
-        // Ensure the header id is well formed.
-        ensure!(self.digest() == self.id(), DagError::InvalidHeaderId);
+        // Ensure the header digest is well formed.
+        ensure!(
+            Hash::digest(self) == self.digest(),
```

Comment:
Excellent, the dispatch priorities between the trait impl and the struct methods will indeed make the method shadow the trait impl.